internal/converter/types.go

```diff
@@ -498,7 +519,57 @@ func (c *Converter) recursiveFindNestedMessages(curPkg *ProtoPackage, msgDesc *d
 	return nil
 }
 
-func (c *Converter) recursiveConvertMessageType(curPkg *ProtoPackage, msgDesc *descriptor.DescriptorProto, pkgName string, duplicatedMessages map[*descriptor.DescriptorProto]string, ignoreDuplicatedMessages bool) (*jsonschema.Type, error) {
+// findNestedMessages takes a message, and returns a map mapping pointers to enum definitions nested within it:
+// these messages become definitions which can be referenced (instead of repeating them every time they're used)
+func (c *Converter) findNestedEnums(curPkg *ProtoPackage, msgDesc *descriptor.DescriptorProto) (map[*descriptor.EnumDescriptorProto]string, error) {
+
+	// Get a list of all nested messages, and how often they occur:
+	nestedEnums := make(map[*descriptor.EnumDescriptorProto]string)
+	if err := c.recursiveFindNestedEnums(curPkg, msgDesc, msgDesc.GetName(), nestedEnums, make(map[*descriptor.DescriptorProto]bool)); err != nil {
+		return nil, err
+	}
+
+	// Now filter them:
+	result := make(map[*descriptor.EnumDescriptorProto]string)
+	for message, enumName := range nestedEnums {
+		if !strings.HasPrefix(enumName, ".google.protobuf.") {
+			result[message] = strings.TrimLeft(enumName, ".")
+		}
+	}
+
+	return result, nil
+}
+
+func (c *Converter) recursiveFindNestedEnums(curPkg *ProtoPackage, msgDesc *descriptor.DescriptorProto, typeName string, nestedEnums map[*descriptor.EnumDescriptorProto]string, searchedMsgs map[*descriptor.DescriptorProto]bool) error {
```

Comment:
added searchedMsgs, a map functioning as a set to handle cyclical references (found bug because of the cyclicalreference test)